### ISSUE TITLE
Suggestions for a more concise ADT API

### ISSUE THREAD

**Issue by boris-marinov** [origamitower/folktale#9]
For starters I want to say that I really like the Algebraic Datatype API - I am looking forward to start using it. 

However the boilerplate involved in defining a type is more than its needed and that can make it confusing for newcomers. Spent some time and I came up with two suggestions for improving (simplifying) the API. 

The two suggestions would probably work best together, but still they can be discussed separately.

## Suggestion one: Passing constructors to the `data` method when defining a variant.

Currently you define a variant by defining its fields (as an array) and the fields are used to create a default constructor for this variant. 

```javascript
const Maybe = data({
  Nothing:[],
  Just:['value']
})
```
I suggest that the user should be able to pass the constructor directly. Seems more concise and gives him/her more possibility:

```javascript
const Maybe = data({
  Nothing: () => ({})
  Just: (value) => ({ value })
})
```
A caveat here is that we are using information - the fields of the datatype will be not specified explicitly. However the user specifies a constructor which we assume does the needed field validation. Not sure what are the implications will be for the `deriving` feature.

## Suggestion two: Returning constructors from the `data` method.

Currently the `data` constructor returns a dictionary of intermediate objects containing various information about each variant:

    interface Variant {
      name        : String,
      constructor : Constructor,
      prototype   : Object,
      fields      : [String]
    }

I think that nothing stops this object from also being the variant's constructor. More so when one of its properties is `prototype`.

## Conclusion

With those two ideas implemented `data` can be perceived as a method that does not generate any structure by itself but rather it augments the structure that you provide it with. For me this feels very intuitive to work with and it has zero boilerplate:

```javascript
const Maybe = data({
  Nothing: () => ({})
  Just: (value) => ({ value })
})

const {Just, Nothing} = Maybe

Maybe.of = Just
Maybe.fromNull = (val) => val == null ? Nothing() : Just(val)
Nothing.prototype.map = function(f) { return this }
Just.prototype.map = function(f) { return f(this.value) }

module.exports = Maybe

```
I am sure that there are some caveats in this which I am overlooking. I will be happy to learn more about them and to continue this conversation.

**Comment by robotlolita** [comment]
Ah, I'm all for making the adt module easier to use. And this is a promising direction :)

TL;DR: we should definitely return more usable constructors out of the box for `data`. But I'd like to keep automatic derivation if at all possible. So far, I believe something like the following could be an improvement, but it doesn't handle validation/etc:

```js
const List = data({
  Nil: [],
  Cons: ['value', 'rest']
}).deriving(Setoid);

const { Nil, Cons } = List;

// .equals provided by the Setoid derivation
Cons(1, Cons(2, Nil())).equals(Cons(1, Cons(2, Nil())))
// => true
```

- - -

## On suggestion one

The current design for `data` takes an array of field names to support reflection and derivation. My initial idea was to have something like:

```js
const Maybe = data({
  Nothing: {},
  Just: { value: isObject }
})
```

Where the key is the name of the field, and the value is a function to validate or transform the value provided to the field. This is similar to how the adt-simple (https://github.com/natefaubion/adt-simple#basic-usage) macro works. The problem here is that until ES2015 the order of properties in objects were not guaranteed, so if you have something like:

```js
const Tuple2 = data({
  Tuple: { left: isObject, right: isObject }
})
```

The keys might be handled as either `[right, left]` or `[left, right]`. Which is... not a problem for how I ended up designing it anyway, given that the constructor takes an object.

Putting the constructor directly in the object you pass to data is a powerful thing, but we lose the reflection capability, which allows sharing some reusable implementations of common things, so you could have:

```js
// This is okay because the property names aren't important
// so long as they are consistent
const Maybe = data({
  Nothing(){ },
  Just(value){ this.value = value }
}).deriving(Setoid);
Maybe.Just(1).equals(Maybe.Just(1))

// This is not okay because the property names are important
const Maybe = data({
  Nothing(){ },
  Just(value){ this.value = value }
}).deriving(ToJSON)
Maybe.Just(1).toJSON()
// (could be any of the following depending on minification)
// => { "value": 1 }
// => { "a": 1 }

// This is worse because it's a function on Maybe itself
const Maybe = data({
  Nothing(){ },
  Just(value){ this.value = value }
}).deriving(FromJSON)
Maybe.FromJSON({ "type": "Just", "value": 1 })
// => ??? how to map 'value' to whatever Just accepts?
```

These automatic derivations will be slow because they use runtime meta-programming, but I think it's a reasonable price to pay in simpler data structures where you don't care about performance too much.

## On suggestion two

Ah, `data` already returns constructors, so something like:

```js
const List = data({ Nil: [], Cons: ['value', 'rest'] });
```

Is equivalent to:

```js
const List = {};
List.Nil = function(){ };
List.Cons = function({ value, rest }) {
  this.value = value;
  this.rest = rest;
}
```

The major problem is that, right now, it requires you to pass it an object and use `new`, so you get `new List.Cons({ value: 1, rest: new List.Nil({}))`. The object is still necessary even if the constructor doesn't take any parameters. Suggestion one could help fix this, but we need a solution for reflection too :(

Instead of returning something that requires `new`, we could return a function that always constructs a new object, so you could have it as: `List.Cons(...)`. and `List.Nil()`, so the `data` would make the wrappers automatically instead of requiring you to do that.

**Comment by boris-marinov** [comment]
Thanks for the good response.

I suspected that the derivations are the missing piece of the puzzle.

Let's go through the examples: 

The **first** one is OK, I think. The only difference between the two setoid implementations (the one which relies on the `fields` array and the one which doesn't) is that one will use `Object.keys`. Intuitively a derived `equals` method (or any derived method for that matter) has to work on objects with any set of properties anyways.

In your **second example** I Could not simulate the issue with the minification: `Just(value){ this.value = value }` gets compiled to `Just(a){ this.value = a }` but that does not bother us.

The **third** one is interesting, but maybe its too early to discuss it.

I think I will work on some experimental derivations which I will test them using the `Either` datatype.

Also `data` [does not return constructors directly](https://github.com/origamitower/folktale/blob/master/src/core/adt/index.js#L69) in the current master (or so it seems) although it looks like the Maybe module [relies on them](https://github.com/origamitower/folktale/blob/master/src/data/maybe/core.js#L70).

**Comment by robotlolita** [comment]
Ah, you're quite right about the first and second example. The only minifier that'll change property names is Google Closure Compiler with the advanced optimisations mode, but that one isn't quite JavaScript, so I don't think we should support it anyway.

Oh, that's a bug in the `data` implementation. There should be an `Object.assign` there :3

**Comment by robotlolita** [comment]
So, considering #10 and your messages, I suppose the ADT module should provide the following characteristics:

  - An easier way of constructing values:
  - Input validation for fields in the constructor.
  - Type name reflection (for error messages, serialisation, etc)
  - Field reflection (for extracting information and constructing values)

So, I'm thinking about the following API:

#### Basic usage

```js
// -- Introduction
const Maybe = data('Maybe', {
  Nothing: [],
  Just: ['value']
});

// -- Elimination
Maybe.Nothing();  // Nothing {}
Maybe.Just(1);    // Just { value: 1 }
```

#### Validations

```js
const isString = (x) => {
  if (typeof x !== "string") {
    throw new TypeError(`Expected a number, got ${x}`)
  }
  return x;
};

// -- Introduction
const Contact = data('Contact', {
  Email: [{ email: isString }],
  Phone: [{ phone: isString }]
});

// -- Elimination
Contact.Email("foo") // Email { email: "foo" }
Contact.Email(null)  // TypeError
Contact.Phone("foo") // Phone { phone: "foo" }
Contact.Phone(222)   // TypeError
```

Recursive structures:

```js
// -- Introduction
const { recursive } = require('folktale/core/adt');
const List = data('List', {
  Nil: [],
  Cons: ['value', { rest: recursive(_ => List.isList) }]
})

// -- Elimination
List.Nil()               // Nil {}
List.Cons(1, 2)          // TypeError
List.Cons(1, List.Nil()) // Cons { value: 1, rest: Nil {} }
```

#### Normalisation

```js
const toUpperCase = (x) => x.toUpperCase()

// -- Introduction
const { Person } = data('Person', {
  Person: [{ name: toUpperCase }]
});

Person("alice") // Person { name: "ALICE" }
```

#### Type name and field reflection

These are reified in the same way they are now, as properties of the Variant. The Variant becomes a function that constructs the value of that type.

**Comment by robotlolita** [comment]
I've started working on this in https://github.com/origamitower/folktale/commit/92af168cc01d904256fb2239ab622082e1c943b7, it's missing the support for recursive types (and it's missing tests), but the other things are implemented.

**Comment by boris-marinov** [comment]
I like the ideas. The syntax still seems a bit non-idiomatic to me, though. Particularily this part `Person: [{ name: isString}, { address: isAddress }]`. I think I am still leaning more towards my original proposition (after all, derivations are just a way to reduce boilerplate, so they shouldn't introduce more of it). But then again I may be biased :)

Recursive types are an interesting aspect of this. Related to them, maybe each type should have a generic method for checking whether a value matches it. So instead of `List.isList` we can have something like `List.instanceOf` (this was just the first name that popped into my mind). I would have used this when implementing the Setoid derivation.

**Comment by boris-marinov** [comment]
Implemented:
https://github.com/origamitower/folktale/commit/ec4126396abb85d95ee4279634a314c9807fcaf1
